packages/contracts/contracts/libs/v0.8.x/ImmutableStringArray.sol

```diff
@@ -0,0 +1,282 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+// Created By: Art Blocks Inc.
+
+pragma solidity ^0.8.0;
+
+import {SSTORE2} from "./SSTORE2.sol"; // Import SSTORE2 library
+
+/**
+ * @title ImmutableStringArray
+ * @author Art Blocks Inc.
+ * @notice This library is optimized to store immutable arrays of strings much more efficiently than using native Solidity storage arrays.
+ * It uses a single SSTORE2 contract to store the length + packed offsets + packed string data.
+ * Allows overwriting the pointer to point to a new immutable array.
+ */
+library ImmutableStringArray {
+    /**
+     * @notice Struct to store the SSTORE2 pointer to the packed string array.
+     * @dev dataPointer is the SSTORE2 pointer to the packed string array.
+     */
+    struct StringArray {
+        address dataPointer; // SSTORE2 pointer storing length + packed offsets + packed string data
+    }
+
+    /**
+     * @notice Stores a packed immutable string array using a single SSTORE2 contract.
+     * Allows overwriting the pointer to point to a new immutable array.
+     * @param storageArray The storage reference to store the packed array.
+     * @param strings The array of strings to pack.
+     */
+    function store(
+        StringArray storage storageArray,
+        string[] memory strings
+    ) internal {
+        uint256 arrayLength = strings.length;
+        uint64[] memory offsets = new uint64[](arrayLength);
+
+        // compute total bytes length and offsets (ensuring no overflow)
+        uint256 totalStringBytesLength;
+        for (uint256 i = 0; i < arrayLength; i++) {
+            offsets[i] = uint64(totalStringBytesLength);
+            totalStringBytesLength += bytes(strings[i]).length;
+        }
+        // realistically will not overflow uint64, but check for security guarantees
+        // @dev no coverage on else (difficulty of triggering)
+        require(
+            totalStringBytesLength <= type(uint64).max,
+            "Offset exceeds uint64 limit"
+        );
+
+        // prepare the combined storage structure (length + packed offsets + packed strings)
+        // @dev 8 bytes for length, 8 bytes per offset, totalStringBytesLength bytes for strings
+        // over-allocate 32 bytes for memory safety
+        // note: shorten length by 32 bytes at end of function before returning
+        uint256 packedDataLength = 8 +
+            (arrayLength * 8) +
+            totalStringBytesLength;
+        bytes memory packedData = new bytes(packedDataLength + 0x20);
+        uint256 ptr;
+
+        // store the length of the strings array in the first 8 bytes of the packedData
+        assembly ("memory-safe") {
+            ptr := add(packedData, 0x20) // pointer to first byte of packedData
+            let existing := mload(ptr) // load the full 32-byte word
```

Comment:
Do we need this?

Reply:
good catch - removed in [4cc824e](https://github.com/ArtBlocks/artblocks-contracts/pull/1767/commits/4cc824e9a82d0e850c23e767e52f65ed6856196f)